README.md
# libp2p-websocket-star

[![](https://img.shields.io/badge/made%20by-mkg20001-blue.svg?style=flat-square)](http://ipn.io)
[![Build Status](https://travis-ci.org/libp2p/js-libp2p-websocket-star.svg?style=flat-square)](https://travis-ci.org/libp2p/js-libp2p-websocket-star)

![](https://raw.githubusercontent.com/libp2p/interface-connection/master/img/badge.png)
![](https://raw.githubusercontent.com/libp2p/interface-transport/master/img/badge.png)

> Allows to listen on multiple websocket-star-rendezvous servers while ignoring offline ones

## Description

`libp2p-websocket-star-multi` allows to listen on multiple websocket-star-rendezvous servers while ignoring offline ones

**Note:** This module uses [pull-streams](https://pull-stream.github.io) for all stream based interfaces.

## Usage

### Installation

```bash
> npm install libp2p-websocket-star
```

### API

[![](https://raw.githubusercontent.com/libp2p/interface-transport/master/img/badge.png)](https://github.com/libp2p/interface-transport)

### Example

```js
'use strict'

const Libp2p = require('libp2p')
const Id = require('peer-id')
const Info = require('peer-info')
const multiaddr = require('multiaddr')
const pull = require('pull-stream')

const WSStarMulti = require('libp2p-websocket-star-multi')

Id.create((err, id) => {
  if (err) throw err

  const peerInfo = new Info(id)
  peerInfo.multiaddrs.add(multiaddr('/p2p-websocket-star')) // will get replaced to the multiaddr of the individual servers
  const ws = new WSStarMulti({
    servers: [ // servers are Multiaddr[]
      '/dnsaddr/ws-star-signal-1.servep2p.com/tcp/443/wss/p2p-websocket-star',
      '/dnsaddr/ws-star-signal-2.servep2p.com/tcp/443/wss/p2p-websocket-star',
      '/dnsaddr/ws-star-signal-3.servep2p.com/tcp/443/wss/p2p-websocket-star',
      '/dnsaddr/ws-star-signal-4.servep2p.com/tcp/443/wss/p2p-websocket-star',
      '/dnsaddr/ws-star.discovery.libp2p.io/tcp/443/wss/p2p-websocket-star',
      '/dns4/localhost/tcp/80/ws/p2p-websocket-star'
    ],
    // ignore_no_online: true, // enable this to prevent wstar-multi from returning a listen error if no servers are online
    id // the id is required for the crypto challenge
  })

  const modules = {
    transport: [
      ws
    ],
    discovery: [
      ws.discovery
    ]
  }

  const node = new Libp2p(modules, peerInfo)

  node.handle('/test/1.0.0', (protocol, conn) => {
    pull(
      pull.values(['hello']),
      conn,
      pull.map((s) => s.toString()),
      pull.log()
    )
  })

  node.start((err) => {
    if (err) {
      throw err
    }

    node.dial(peerInfo, '/test/1.0.0', (err, conn) => {
      if (err) {
        throw err
      }

      pull(
        pull.values(['hello from the other side']),
        conn,
        pull.map((s) => s.toString()),
        pull.log()
      )
    })
  })
})
```

Outputs:
```
hello
hello from the other side
```

### This module uses `pull-streams`

We expose a streaming interface based on `pull-streams`, rather then on the Node.js core streams implementation (aka Node.js streams). `pull-streams` offers us a better mechanism for error handling and flow control guarantees. If you would like to know more about why we did this, see the discussion at this [issue](https://github.com/ipfs/js-ipfs/issues/362).

You can learn more about pull-streams at:

- [The history of Node.js streams, nodebp April 2014](https://www.youtube.com/watch?v=g5ewQEuXjsQ)
- [The history of streams, 2016](http://dominictarr.com/post/145135293917/history-of-streams)
- [pull-streams, the simple streaming primitive](http://dominictarr.com/post/149248845122/pull-streams-pull-streams-are-a-very-simple)
- [pull-streams documentation](https://pull-stream.github.io/)

#### Converting `pull-streams` to Node.js Streams

If you are a Node.js streams user, you can convert a pull-stream to a Node.js stream using the module [`pull-stream-to-stream`](https://github.com/pull-stream/pull-stream-to-stream), giving you an instance of a Node.js stream that is linked to the pull-stream. For example:

```js
const pullToStream = require('pull-stream-to-stream')

const nodeStreamInstance = pullToStream(pullStreamInstance)
// nodeStreamInstance is an instance of a Node.js Stream
```

To learn more about this utility, visit https://pull-stream.github.io/#pull-stream-to-stream.

LICENSE MIT


CHANGELOG.md
<a name="0.4.4"></a>
## [0.4.4](https://github.com/mkg20001/js-libp2p-websocket-star-multi/compare/v0.4.3...v0.4.4) (2019-03-19)


### Bug Fixes

* **test:** travis ipv6 fix ([ed1bff0](https://github.com/mkg20001/js-libp2p-websocket-star-multi/commit/ed1bff0))
* reduce bundle size ([edd3a4b](https://github.com/mkg20001/js-libp2p-websocket-star-multi/commit/edd3a4b))



<a name="0.4.3"></a>
## [0.4.3](https://github.com/mkg20001/js-libp2p-websocket-star-multi/compare/v0.4.0...v0.4.3) (2019-01-24)



<a name="0.4.2"></a>
## [0.4.2](https://github.com/mkg20001/js-libp2p-websocket-star-multi/compare/v0.4.0...v0.4.2) (2019-01-24)



<a name="0.4.1"></a>
## [0.4.1](https://github.com/mkg20001/js-libp2p-websocket-star-multi/compare/v0.4.0...v0.4.1) (2019-01-24)



<a name="0.4.0"></a>
# [0.4.0](https://github.com/mkg20001/js-libp2p-websocket-star-multi/compare/v0.3.2...v0.4.0) (2018-10-23)


### Bug Fixes

* fix multiaddr bug, update packages ([2898286](https://github.com/mkg20001/js-libp2p-websocket-star-multi/commit/2898286))
* lint ([16e8012](https://github.com/mkg20001/js-libp2p-websocket-star-multi/commit/16e8012))



<a name="0.3.2"></a>
## [0.3.2](https://github.com/mkg20001/js-libp2p-websocket-star-multi/compare/v0.3.1...v0.3.2) (2018-02-12)


### Bug Fixes

* /dns{ => addr} in readme ([1ca38b1](https://github.com/mkg20001/js-libp2p-websocket-star-multi/commit/1ca38b1))



<a name="0.3.1"></a>
## [0.3.1](https://github.com/mkg20001/js-libp2p-websocket-star-multi/compare/v0.3.0...v0.3.1) (2018-02-12)


### Bug Fixes

* Update to latest ws-star to fix issues ([8b57068](https://github.com/mkg20001/js-libp2p-websocket-star-multi/commit/8b57068))



<a name="0.3.0"></a>
# [0.3.0](https://github.com/mkg20001/js-libp2p-websocket-star-multi/compare/v0.2.1...v0.3.0) (2017-12-07)


### Features

* Use better codestyle and new ws-star ([aecb2a6](https://github.com/mkg20001/js-libp2p-websocket-star-multi/commit/aecb2a6))



<a name="0.2.1"></a>
## [0.2.1](https://github.com/mkg20001/js-libp2p-websocket-star-multi/compare/v0.2.0...v0.2.1) (2017-10-28)



<a name="0.2.0"></a>
# [0.2.0](https://github.com/mkg20001/js-libp2p-websocket-star-multi/compare/v0.1.1...v0.2.0) (2017-09-26)



<a name="0.1.1"></a>
## [0.1.1](https://github.com/mkg20001/js-libp2p-websocket-star-multi/compare/v0.1.0...v0.1.1) (2017-09-02)



<a name="0.1.0"></a>
# 0.1.0 (2017-08-31)





test/listen-and-dial.spec.js
/* eslint-env mocha */
/* eslint-disable max-nested-callbacks */

'use strict'

const servers = ['/ip4/127.0.0.1/tcp/15001/ws/p2p-websocket-star', '/ip4/127.0.0.1/tcp/15002/ws/p2p-websocket-star'].concat(!process.env.TRAVIS ? ['/ip6/::1/tcp/15003/ws/p2p-websocket-star'] : [])
const offlineServer = '/ip4/127.0.0.1/tcp/16000/ws/p2p-websocket-star'

const chai = require('chai')
const dirtyChai = require('dirty-chai')
const expect = chai.expect
const multiaddr = require('multiaddr')
const getMa = id => multiaddr('/p2p-websocket-star/ipfs/' + id.toB58String())
const pull = require('pull-stream')
const pullTestData = [Buffer.from('hello world!')]
const pullTest = (conn, cb) => pull(pull.values(pullTestData.slice(0)), conn, pull.collect((err, res) => {
  expect(err).to.not.exist()
  expect(res).to.eql(pullTestData)
  cb()
}))
chai.use(dirtyChai)

const Id = require('peer-id')
const {
  each,
  map
} = require('async')
const WSMulti = require('..')

describe('websocket-star-multi', () => {
  let ids
  before(cb => map(require('./ids.json'), Id.createFromJSON, (err, _) => err ? cb(err) : cb(null, (ids = _))))
  let l = []
  describe('listen', () => {
    let id
    before(() => {
      id = ids[0]
    })
    it('should listen on 1 address', cb => {
      const listener = new WSMulti({
        id,
        servers: [servers[0]]
      }).createListener()
      l.push(listener)
      listener.listen(getMa(id), err => {
        expect(err).to.not.exist()
        expect(listener.online).to.have.lengthOf(1)
        cb()
      })
    })

    it('should listen on 3 addresses', cb => {
      const listener = new WSMulti({
        id,
        servers: servers
      }).createListener()
      l.push(listener)
      listener.listen(getMa(id), err => {
        expect(err).to.not.exist()
        expect(listener.online).to.have.lengthOf(!process.env.TRAVIS ? 3 : 2)
        cb()
      })
    })

    it('listen on offline server should fail', cb => {
      const listener = new WSMulti({
        id,
        servers: [offlineServer]
      }).createListener()
      l.push(listener)
      listener.once('error', () => {})
      listener.listen(getMa(id), err => {
        expect(err).to.exist()
        expect(listener.online).to.have.lengthOf(0)
        cb()
      })
    })

    it('listen on offline server and up server should not fail', cb => {
      const listener = new WSMulti({
        id,
        servers: [servers[0], offlineServer]
      }).createListener()
      l.push(listener)
      listener.listen(getMa(id), err => {
        expect(err).to.not.exist()
        expect(listener.online).to.have.lengthOf(1)
        cb()
      })
    })

    it('listen on offline server with ignore_no_online should not fail', cb => {
      const listener = new WSMulti({
        id,
        servers: [offlineServer],
        ignore_no_online: true
      }).createListener()
      l.push(listener)
      listener.listen(getMa(id), err => {
        expect(err).to.not.exist()
        expect(listener.online).to.have.lengthOf(0)
        cb()
      })
    })
  })
  describe('dial', () => {
    let c1, c2, c3
    before(cb => {
      map(ids, (id, cb) => {
        const m = new WSMulti({
          id,
          servers: servers
        })
        const l = m.createListener(conn => pull(conn, conn))
        l.id = id
        l.dial = m.dial.bind(m)
        l.listen(getMa(id), err => err ? cb(err) : cb(null, l))
      }, (err, c) => {
        expect(err).to.not.exist()
        c1 = c.shift()
        c2 = c.shift()
        c3 = c.shift()
        cb()
      })
    })

    it('c1 should dial to c2 over server 2', cb => {
      const toDial = multiaddr(servers[1] + '/ipfs/' + c2.id.toB58String())
      c1.dial(toDial, (err, conn) => {
        expect(err).to.not.exist()
        pullTest(conn, cb)
      })
    })

    it('c1 dial c3 over non-existent server should fail', cb => {
      const toDial = multiaddr(offlineServer + '/ipfs/' + c2.id.toB58String())
      c1.dial(toDial, (err, conn) => {
        expect(err).to.exist()
        expect(conn).to.not.exist()
        cb()
      })
    })

    after(() => l.push(c1, c2, c3))
  })
  afterEach(cb => each(l, (l, cb) => l.close(cb), err => err ? cb(err) : cb(null, (l = []))))
})


gulpfile.js
'use strict'

const gulp = require('gulp')
const parallel = require('async/parallel')
const rendezvous = require('libp2p-websocket-star-rendezvous')

let _r = []

gulp.task('test:node:before', boot)
gulp.task('test:node:after', stop)
gulp.task('test:browser:before', boot)
gulp.task('test:browser:after', stop)

function boot (done) {
  const base = (v) => Object.assign({
    host: '0.0.0.0',
    refreshPeerListIntervalMS: 1000
  }, v)

  parallel([['r1', {port: 15001}], ['r2', {port: 15002}], ['r3', {port: 15003, host: '::'}]].map((v) => (cb) => {
    rendezvous.start(base(v.pop()), (err, r) => {
      if (err) { return cb(err) }
      _r.push(r)
      console.log('%s: %s', v.pop(), r.info.uri)
      cb()
    })
  }), done)
}

function stop (done) {
  parallel(_r.map((r) => (cb) => r.stop(cb)), done)
  _r = []
}

require('aegir/gulp')(gulp)


src/index.js
'use strict'

const debug = require('debug')
const log = debug('libp2p:websocket-star:multi')
const once = require('once')

const EE = require('events').EventEmitter
const map = require('async/map')
const parallel = require('async/parallel')
const multiaddr = require('multiaddr')
const mafmt = require('mafmt')

const WSStar = require('libp2p-websocket-star')
class WebsocketStarMulti { // listen on multiple websocket star servers without having to worry about one being down.
  // NOTE: if no servers are reachable or provided an error is thrown
  constructor (opt) {
    this.opt = opt || {}
    this.servers = opt.servers || []
    this.ws = new WSStar(this.opt)
    this.discovery = this.ws.discovery
  }
  dial (ma, opt, cb) {
    log('dial', ma)
    return this.ws.dial(ma, opt, cb)
  }
  createListener (options, handler) {
    if (typeof options === 'function') {
      handler = options
      options = {}
    }

    const listener = new EE()
    listener.servers = {}
    listener.online = []
    this.servers.forEach(ser => {
      const s = this.ws.createListener(options, handler)
      s.once('error', () => {})
      s.url = ser
      listener.servers[ser] = s
    })

    listener.listen = (ma, cb) => {
      const id = ma.toString().split('ipfs/').pop()
      log('listen on %s server(s) with id %s', this.servers.length, id)
      parallel(this.servers.map(url => listener.servers[url]).map(server =>
        cb => {
          log('listen %s', server.url)
          const next = once(err => {
            log('listen %s ok %s', server.url, !err)
            if (err) return cb(log(err))
            listener.online.push(server)
            return cb()
          })
          setTimeout(next, this.opt.timeout || 5000, new Error('Timeout'))
          server.listen(multiaddr(server.url).encapsulate('/ipfs/' + id), next)
        }), () => {
        if (!listener.online.length && !this.opt.ignore_no_online) {
          const e = new Error("Couldn't listen on any of the servers")
          listener.emit('error', e)
          cb(e)
        } else {
          listener.emit('listening')
          cb()
        }
      })
    }

    listener.close = cb =>
      parallel(listener.online.map(s => cb => s.close(cb)), err => cb(err, (listener.online = [])))

    listener.getAddrs = cb => map(listener.online, (s, n) => s.getAddrs(n), (err, res) => {
      if (err) return cb(err)
      return cb(null, res.reduce((a, b) => a.concat(b), []))
    })

    return listener
  }

  filter (ma) {
    if (!Array.isArray(ma)) ma = [ma]
    return ma.filter(ma => ma.toString().startsWith('/p2p-websocket-star') || mafmt.WebSocketStar.matches(ma))
  }
}

module.exports = WebsocketStarMulti


example.js
'use strict'

const Libp2p = require('libp2p')
const Id = require('peer-id')
const Info = require('peer-info')
const multiaddr = require('multiaddr')
const pull = require('pull-stream')

const WSStarMulti = require('libp2p-websocket-star-multi')

Id.create((err, id) => {
  if (err) throw err

  const peerInfo = new Info(id)
  peerInfo.multiaddrs.add(multiaddr('/p2p-websocket-star')) // will get replaced to the multiaddr of the individual servers
  const ws = new WSStarMulti({
    servers: [ // servers are Multiaddr[]
      '/dnsaddr/ws-star-signal-1.servep2p.com/tcp/443/wss/p2p-websocket-star',
      '/dnsaddr/ws-star-signal-2.servep2p.com/tcp/443/wss/p2p-websocket-star',
      '/dnsaddr/ws-star-signal-3.servep2p.com/tcp/443/wss/p2p-websocket-star',
      '/dnsaddr/ws-star-signal-4.servep2p.com/tcp/443/wss/p2p-websocket-star',
      '/dnsaddr/ws-star.discovery.libp2p.io/tcp/443/wss/p2p-websocket-star',
      '/dns4/localhost/tcp/80/ws/p2p-websocket-star'
    ],
    // ignore_no_online: true, // enable this to prevent wstar-multi from returning a listen error if no servers are online
    id // the id is required for the crypto challenge
  })

  const modules = {
    transport: [
      ws
    ],
    discovery: [
      ws.discovery
    ]
  }

  const node = new Libp2p(modules, peerInfo)

  node.handle('/test/1.0.0', (protocol, conn) => {
    pull(
      pull.values(['hello']),
      conn,
      pull.map((s) => s.toString()),
      pull.log()
    )
  })

  node.start((err) => {
    if (err) {
      throw err
    }

    node.dial(peerInfo, '/test/1.0.0', (err, conn) => {
      if (err) {
        throw err
      }

      pull(
        pull.values(['hello from the other side']),
        conn,
        pull.map((s) => s.toString()),
        pull.log()
      )
    })
  })
})
